File: validate-binary-search-tree/validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        
        
        def dfs(root, curr_min, curr_max):
            
            if not root:
                return True
            
            if root.val <= curr_min or root.val >= curr_max:
                return False
            
            return dfs(root.left,curr_min, root.val) and dfs(root.right, root.val, curr_max)
        
        return dfs(root, float('-inf'), float('inf'))

File: validate-binary-search-tree/validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        if (root == null) return true;
        return dfs(root, null, null);
    }
    
    public static boolean dfs(TreeNode root, Integer max_val, Integer min_val){
        
        if(root == null)return true;
        
        if ((min_val != null && root.val <= min_val) || (max_val != null && root.val >= max_val)) return false;
        
        
        return dfs(root.left, root.val, min_val) && dfs(root.right, max_val, root.val); 
        
    } 
}